hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/KeyDeletingService.java

```diff
@@ -186,9 +173,116 @@ public void setKeyLimitPerTask(int keyLimitPerTask) {
+      try {
+        // TODO: [SNAPSHOT] HDDS-7968. Reclaim eligible key blocks in
+        //  snapshot's deletedTable when active DB's deletedTable
+        //  doesn't have enough entries left.
+        //  OM would have to keep track of which snapshot the key is coming
+        //  from if the above would be done inside getPendingDeletionKeys().
+        OmSnapshotManager omSnapshotManager = getOzoneManager().getOmSnapshotManager();
+        // This is to avoid race condition b/w purge request and snapshot chain update. For AOS taking the global
+        // snapshotId since AOS could process multiple buckets in one iteration. While using path
+        // previous snapshotId for a snapshot since it would process only one bucket.
+        UUID expectedPreviousSnapshotId = currentSnapshotInfo == null ?
+            snapshotChainManager.getLatestGlobalSnapshotId() :
+            SnapshotUtils.getPreviousSnapshotId(currentSnapshotInfo, snapshotChainManager);
+
+        IOzoneManagerLock lock = getOzoneManager().getMetadataManager().getLock();
+
+        // Purge deleted Keys in the deletedTable && rename entries in the snapshotRenamedTable which doesn't have a
+        // reference in the previous snapshot.
+        try (ReclaimableKeyFilter reclaimableKeyFilter = new ReclaimableKeyFilter(getOzoneManager(),
+            omSnapshotManager, snapshotChainManager, currentSnapshotInfo, keyManager, lock)) {
+          // Get pending keys that can be deleted
+          PendingKeysDeletion pendingKeysDeletion = currentSnapshotInfo == null
+              ? keyManager.getPendingDeletionKeys(reclaimableKeyFilter, remainNum)
+              : keyManager.getPendingDeletionKeys(volume, bucket, startKey, reclaimableKeyFilter, remainNum);
+          List<BlockGroup> keyBlocksList = pendingKeysDeletion.getKeyBlocksList();
+          //submit purge requests if there are renamed entries to be purged or keys to be purged.
+          if (keyBlocksList != null && !keyBlocksList.isEmpty()) {
+            // Validating if the previous snapshot is still the same before purging the blocks.
+            SnapshotUtils.validatePreviousSnapshotId(currentSnapshotInfo, snapshotChainManager,
+                expectedPreviousSnapshotId);
+            Pair<Integer, Boolean> purgeResult = processKeyDeletes(keyBlocksList,
+                 pendingKeysDeletion.getKeysToModify(), snapshotTableKey,
+                 expectedPreviousSnapshotId);
+            remainNum -= purgeResult.getKey();
+            successStatus = purgeResult.getValue();
+            metrics.incrNumKeysProcessed(keyBlocksList.size());
+            metrics.incrNumKeysSentForPurge(purgeResult.getKey());
+            if (successStatus) {
+              deletedKeyCount.addAndGet(purgeResult.getKey());
+            }
+          }
+
+          // Checking remainNum is greater than zero and not equal to the initial value if there were some keys to
+          // reclaim. This is to check if
```

Comment:
Incomplete comment?

Reply:
done updated it. Missed this while resolving merge conflict